orion-server/src/main/java/com/pinterest/orion/core/actions/kafka/ClusterRecoveryAction.java

```diff
@@ -0,0 +1,223 @@
+        Action brokerRecoveryAction = newBrokerRecoveryAction();
+        brokerRecoveryAction.setAttribute(OrionConstants.NODE_ID, deadBrokerId, sensorSet);
+        brokerRecoveryAction.setOwner("ClusterRecoveryAction");
+
+        if (nonExistentBrokers.contains(deadBrokerId)) {
+            Node existingNode = cluster.getNodeMap().values().iterator().next();
+            String extractedName = deriveNonexistentHostname(
+                    existingNode.getCurrentNodeInfo().getHostname(),
+                    existingNode.getCurrentNodeInfo().getNodeId(),
+                    deadBrokerId
+            );
+            // setting these attributes to indicate that the node doesn't exist in cluster map, and should skip any node-related checks
+            brokerRecoveryAction.setAttribute(BrokerRecoveryAction.ATTR_NODE_EXISTS_KEY, false);
+            brokerRecoveryAction.setAttribute(BrokerRecoveryAction.ATTR_NONEXISTENT_HOST_KEY, extractedName);
+        }
+        if (maybeDeadBrokers.contains(deadBrokerId)) {
+            // Setting this flag in the action will restart the broker before replacing the broker
+            brokerRecoveryAction.setAttribute(BrokerRecoveryAction.ATTR_TRY_TO_RESTART_KEY, true);
+            String restartNote = "Will try to restart node " + deadBrokerId + " before replacing it. ";
+            logger.info(restartNote);
+            getResult().appendOut(restartNote);
+        }
+        String dispatchNote = String.format(
+                "Dispatching BrokerRecoveryAction for node %s in cluster %s. ",
+                deadBrokerId,
+                cluster.getClusterId());
+        logger.info(dispatchNote);
+        getResult().appendOut(dispatchNote);
+        getEngine().dispatch(brokerRecoveryAction);
+    }
+
+    protected boolean healBrokers(Set<String> candidates) throws Exception {
+        String output;
+        boolean isSucceed = false;
+        if (candidates.size() == 1) {
+            output = String.format(
+                    "ClusterRecoveryAction trys to recover broker %s. ",
+                    candidates.iterator().next());
+            String deadBrokerId = candidates.iterator().next();
+            healBroker(deadBrokerId);
+            logger.info(output);
+            isSucceed = true;
+        } else if (candidates.size() > 1){
+            // more than 1 brokers are dead... better alert and have human intervention
+            output = String.format("More than one brokers are in bad state - dead: %s, service down: %s. " +
+                            "ClusterRecoveryAction skips automatic recovery and pages oncall. ",
+                    deadBrokers, maybeDeadBrokers);
+            cluster.getActionEngine().alert(AlertLevel.HIGH, new AlertMessage(
+                    candidates.size() + " brokers on " + cluster.getClusterId() + " are unhealthy",
+                    output,
+                    "orion"
+            ));
```

Comment:
Can we not create direct pagers from code and base alerting on metrics only?

Reply:
The alert is mutable in cluster configurations. This part is just determining the level of alert, not page code itself.